File: frontend/src/hails/hailEffectFieldLayout.ts
/** Hero-centric effect field sizing — imprint-hail-hero-centric-effect-field E1 */

export const TRANSPORTER_FOOTPRINT_FRACTIONS: Record<string, { width: number; height: number }> = {
  compact: { width: 0.3, height: 0.7 },
  standard: { width: 0.58, height: 0.88 },
  dramatic: { width: 0.52, height: 0.94 },
};

const FOOTPRINT_GLYPH_FLOOR: Record<EffectFootprintProfile, { width: number; height: number }> = {
  compact: { width: 1.04, height: 1.2 },
  standard: { width: 1.08, height: 1.3 },
  dramatic: { width: 1.12, height: 1.4 },
};

export const EFFECT_FOOTPRINT_PROFILES = [
  { id: "compact" as const, hint: "Tight column — minimal safe-zone breakout" },
  { id: "standard" as const, hint: "Default hero-centric footprint" },
  { id: "dramatic" as const, hint: "Widest column still clipped at safe zone" },
];

export type EffectFootprintProfile = "compact" | "standard" | "dramatic";

export type EffectFieldRegion = {
  left: number;
  top: number;
  width: number;
  height: number;
  center_x: number;
  center_y: number;
  bottom: number;
  shape: "column" | "rect" | "radial";
  anchor: "glyph_optical_center";
  effect_field_fraction: {
    width_of_safe_zone: number;
    height_of_safe_zone: number;
  };
  effect_footprint_profile: EffectFootprintProfile;
  glyph_optical_center: { x: number; y: number };
};

function clamp(value: number, lo: number, hi: number, fallback: number): number {
  if (!Number.isFinite(value)) {
    return fallback;
  }
  return Math.max(lo, Math.min(hi, value));
}

export function resolveEffectFootprintProfile(value: unknown): EffectFootprintProfile {
  const profile = String(value ?? "standard").trim().toLowerCase();
  if (profile === "compact" || profile === "dramatic" || profile === "standard") {
    return profile;
  }
  return "standard";
}

export const normalizeEffectFootprintProfile = resolveEffectFootprintProfile;

const FOOTPRINT_LABELS: Record<EffectFootprintProfile, string> = {
  compact: "Compact",
  standard: "Standard",
  dramatic: "Dramatic",
};

export function displayEffectFootprintProfile(profile: string): string {
  const resolved = resolveEffectFootprintProfile(profile);
  return FOOTPRINT_LABELS[resolved];
}

export function computeEffectFieldRegion(input: {
  safeZone: { left: number; top: number; width: number; height: number };
  glyphVisualSizePx: number;
  effectId?: string;
  effectFootprintProfile?: string;
  footprintScale?: number;
}): EffectFieldRegion {
  const profile = resolveEffectFootprintProfile(input.effectFootprintProfile);
  const effectId = input.effectId ?? "transporter";
  const fractions =
    effectId === "transporter"
      ? TRANSPORTER_FOOTPRINT_FRACTIONS[profile] ?? TRANSPORTER_FOOTPRINT_FRACTIONS.standard
      : TRANSPORTER_FOOTPRINT_FRACTIONS.standard;
  const scale = clamp(input.footprintScale ?? 1, 0.85, 1.25, 1);

  const opticalCx = input.safeZone.left + input.safeZone.width / 2;
  const opticalCy = input.safeZone.top + input.safeZone.height / 2;

  let fieldW = input.safeZone.width * fractions.width * scale;
  let fieldH = input.safeZone.height * fractions.height * scale;
  const floors = FOOTPRINT_GLYPH_FLOOR[profile];
  fieldW = Math.max(fieldW, input.glyphVisualSizePx * floors.width);
  fieldH = Math.max(fieldH, input.glyphVisualSizePx * floors.height);
  fieldW = Math.min(fieldW, input.safeZone.width);
  fieldH = Math.min(fieldH, input.safeZone.height);

  let left = opticalCx - fieldW / 2;
  let top = opticalCy - fieldH / 2;
  left = Math.max(input.safeZone.left, Math.min(left, input.safeZone.left + input.safeZone.width - fieldW));
  top = Math.max(input.safeZone.top, Math.min(top, input.safeZone.top + input.safeZone.height - fieldH));

  return {
    left,
    top,
    width: fieldW,
    height: fieldH,
    center_x: opticalCx,
    center_y: opticalCy,
    bottom: top + fieldH,
    shape: effectId === "transporter" ? "column" : "rect",
    anchor: "glyph_optical_center",
    effect_field_fraction: {
      width_of_safe_zone: input.safeZone.width ? fieldW / input.safeZone.width : 0,
      height_of_safe_zone: input.safeZone.height ? fieldH / input.safeZone.height : 0,
    },
    effect_footprint_profile: profile,
    glyph_optical_center: { x: opticalCx, y: opticalCy },
  };
}

File: backend/hails/hail_effect_field_layout.py
"""Hero-centric effect field sizing — imprint-hail-hero-centric-effect-field E1."""

from __future__ import annotations

from typing import Any

TRANSPORTER_FOOTPRINT_FRACTIONS: dict[str, tuple[float, float]] = {
    "compact": (0.30, 0.70),
    "standard": (0.58, 0.88),
    "dramatic": (0.52, 0.94),
}

DEFAULT_FOOTPRINT_FRACTIONS: tuple[float, float] = (0.58, 0.88)

# Minimum field size as multiples of glyph visual size — profile-aware so compact ≠ standard.
_FOOTPRINT_GLYPH_FLOOR: dict[str, tuple[float, float]] = {
    "compact": (1.04, 1.20),
    "standard": (1.08, 1.30),
    "dramatic": (1.12, 1.40),
}


def _clamp(value: float, lo: float, hi: float, default: float) -> float:
    if not isinstance(value, (int, float)) or value != value:
        return default
    return max(lo, min(hi, float(value)))


def resolve_effect_footprint_profile(value: Any) -> str:
    profile = str(value or "standard").strip().lower()
    if profile in TRANSPORTER_FOOTPRINT_FRACTIONS:
        return profile
    return "standard"


def footprint_fractions(effect_id: str, profile: str) -> tuple[float, float]:
    if effect_id == "transporter":
        return TRANSPORTER_FOOTPRINT_FRACTIONS.get(profile, DEFAULT_FOOTPRINT_FRACTIONS)
    return DEFAULT_FOOTPRINT_FRACTIONS


def compute_effect_field_region(
    *,
    safe_zone: dict[str, float],
    glyph_visual_size_px: float,
    effect_id: str = "transporter",
    effect_footprint_profile: str = "standard",
    footprint_scale: float = 1.0,
) -> dict[str, Any]:
    """Axis-aligned effect field inside safe_zone, centered on glyph optical center."""
    profile = resolve_effect_footprint_profile(effect_footprint_profile)
    width_frac, height_frac = footprint_fractions(effect_id, profile)
    scale = _clamp(footprint_scale, 0.85, 1.25, 1.0)

    optical_cx = safe_zone["left"] + safe_zone["width"] / 2.0
    optical_cy = safe_zone["top"] + safe_zone["height"] / 2.0

    field_w = safe_zone["width"] * width_frac * scale
    field_h = safe_zone["height"] * height_frac * scale
    width_floor_mul, height_floor_mul = _FOOTPRINT_GLYPH_FLOOR.get(profile, (1.08, 1.30))
    field_w = max(field_w, glyph_visual_size_px * width_floor_mul)
    field_h = max(field_h, glyph_visual_size_px * height_floor_mul)
    field_w = min(field_w, safe_zone["width"])
    field_h = min(field_h, safe_zone["height"])

    left = optical_cx - field_w / 2.0
    top = optical_cy - field_h / 2.0
    left = max(safe_zone["left"], min(left, safe_zone["left"] + safe_zone["width"] - field_w))
    top = max(safe_zone["top"], min(top, safe_zone["top"] + safe_zone["height"] - field_h))

    shape = "column" if effect_id == "transporter" else "rect"
    return {
        "left": left,
        "top": top,
        "width": field_w,
        "height": field_h,
        "center_x": optical_cx,
        "center_y": optical_cy,
        "shape": shape,
        "anchor": "glyph_optical_center",
        "effect_field_fraction": {
            "width_of_safe_zone": field_w / safe_zone["width"] if safe_zone["width"] else 0.0,
            "height_of_safe_zone": field_h / safe_zone["height"] if safe_zone["height"] else 0.0,
        },
        "effect_footprint_profile": profile,
        "glyph_optical_center": {"x": optical_cx, "y": optical_cy},
        "bottom": top + field_h,
    }


def effect_field_within_safe_zone(effect_field: dict[str, Any], safe_zone: dict[str, float]) -> bool:
    left = float(effect_field["left"])
    top = float(effect_field["top"])
    right = left + float(effect_field["width"])
    bottom = top + float(effect_field["height"])
    safe_right = safe_zone["left"] + safe_zone["width"]
    safe_bottom = safe_zone["top"] + safe_zone["height"]
    return (
        left >= safe_zone["left"] - 0.01
        and top >= safe_zone["top"] - 0.01
        and right <= safe_right + 0.01
        and bottom <= safe_bottom + 0.01
    )
